### DISCUSSION TITLE
React 18 and streaming SSR

### DISCUSSION THREAD

**Discussion by artem-malko** [reduxjs/react-redux#1874] [General]
Hi!

As you know, React 18 has the brand new SSR rendering type — streaming with fully suspense support. So, the questions are:

1. How to pass the server state to the client side?
2. How to rehydrate a redux store on the client side?

Do you have any ideas, how to implement it with streaming SSR?

There is a similar discussion in react-query repo — https://github.com/tannerlinsley/react-query/discussions/2942

**Comment by markerikson** [comment]
This is a complicated topic and I don't have a lot of answers myself.

From what I've read, I don't think _streaming_ SSR will really work with Redux.

The basic pattern of rehydration is the same as it's always been: [serialize the Redux state on the server, load it when you create the store on the client](https://redux.js.org/usage/server-rendering).

If you're using RTK Query, we have specific guidelines on how to do [SSR with Next](https://redux-toolkit.js.org/rtk-query/usage/server-side-rendering) and [rehydration](https://redux-toolkit.js.org/rtk-query/usage/persistence-and-rehydration).

For React-Redux v8, there's [a new `serverState` prop for `<Provider>`](https://github.com/reduxjs/react-redux/releases/tag/v8.0.0-beta.2) you can use to ensure the initial render matches what was on the client.

Beyond that, I'd suggest reading these threads:

- https://github.com/reduxjs/redux-toolkit/discussions/1184
- https://github.com/reduxjs/redux-toolkit/discussions/1820
- https://github.com/reactwg/react-18/discussions/121

and I'll poke @ephem to see if he has any further thoughts (although I see he commented in the React Query thread already)

**Reply by artem-malko** [reply]
Thx for the answer.

"Incremental" external store hydration looks quite interesting, I saw this topic,

Actually, I thought you have some insides from React Work Group. But your answer makes sense already.

**Comment by therealgilles** [comment]
Is there an example repo showing how to use the new serverState prop with SSR (in react-redux with react 18 in general)?

@markerikson: " From what I've read, I don't think streaming SSR will really work with Redux. "
Are you still convinced this is the case?

**Reply by markerikson** [reply]
No example repos yet, which is one of the reasons why I want to add a test :)

Loosely put, what you want to do is:

```js
<Provider store={store} serverState={serializedStateFromServerSSR}>
  <App />
</Provider>
```

And yes, while I am in no way an SSR expert, @Ephem is, and his comments make it sound like streaming SSR is just a fundamentally different thing.

**Reply by therealgilles** [reply]
Thanks for the quick reply. Passing the store state as a window variable for the client as before is still necessary, right?

**Reply by markerikson** [reply]
That would be the standard approach I know of, yeah:

https://redux.js.org/usage/server-rendering#the-client-side

**Reply by therealgilles** [reply]
Thanks again. In this [link](https://github.com/reduxjs/react-redux/releases/tag/v8.0.0-beta.2) about the new serverState prop, the suggestion is `<Provider store={store} serverState={window.initialServerState}>`, which seems odd as window does not exist on the server but I'm guessing it is meant as a reference to the value passed to that window variable.

**Reply by markerikson** [reply]
Hmm. Yeah, I guess you'd need to check for `typeof window !== "undefined"` if you're doing Next usage on the server or similar.

**Reply by artem-malko** [reply]
@therealgilles here you are:

- Server side: https://github.com/artem-malko/react-ssr-template/blob/main/src/framework/applications/server/createApplicationRouteHandler.tsx
- Client side: https://github.com/artem-malko/react-ssr-template/blob/main/src/framework/applications/client/index.tsx

And there is a demo: http://174.138.13.187:5000/news?p=2

By the way, @markerikson , it looks quite strange, that we have to pass initial state from server to `Provider` and to `createStore` https://github.com/artem-malko/react-ssr-template/tree/main/src/framework/applications/client/store/index.ts#L28 Is it ok? Looks like I'm doing something wrong.

**Reply by markerikson** [reply]
That's intentional as far as I know.  

React-Redux uses whatever the "current" state is as it's rendering a component, ie, `store.getState()`.  That _can_ be updated during the rendering process, especially if that render is interrupted.

By passing that state explicitly down through `Provider` to `useSyncExternalStore`, React knows to use that value through the entire initial hydration pass, regardless of what changes may occur to the "current" state during that time.

@acdlite can talk about the intent there more.

**Reply by therealgilles** [reply]
@artem-malko: Thank you for the code examples, very helpful.

I have a remaining question. I use redux-logic, which requires an async call to store.logicMiddleware.whenComplete(). See this [open issue](https://github.com/jeffbski/redux-logic/issues/139). Wondering how to do this with renderToPipeableStream().

```JS
// create store
const store = ...

// maybe dispatch some action here
store.dispatch(...)

// html + head + open body
const html = '<html><head>...</head><body>'

// open react container div
html += '<div id="root">'

// render
html += renderToString(<AppWithProviders store={store} ... />)

// assuming you have put your logicMiddleware on the store when you created it,
// the promise completes when everything in redux-logic is done running
await store.logicMiddleware.whenComplete()
// now I am ready to deliver for SSR

// close react container div
html += '</div>'

// add store preloaded state to html
html += `<script>window.__PRELOADED_STATE__ = ${store.getState()};</script>`

// close body + html tags
html += '</body></html>'

// send full page html
res.status(200).send(html)
```

**Reply by markerikson** [reply]
Hmm. A couple side questions:

- What kind of behavior are you using `redux-logic` for?
- Have you considered using [the new RTK "listener" middleware](https://redux-toolkit.js.org/api/createListenerMiddleware) instead?

I'm actually curious how you would go about doing something similar with listeners here

**Reply by therealgilles** [reply]
- I use redux-logic to handle websockets, REST API (async) calls, and debouncing a string search input field.
- I've quickly looked at RTK and I guess I'm not ready to make the switch yet.

My app does not really need to the whenComplete() call as I don't dispatch any actions or fetch anything on the server side. But as mentioned in the open issue, I'm sure there are valid use cases.

**Reply by therealgilles** [reply]
I updated the code above to better show how the full document html was built before.

**Reply by Ephem** [reply]
Two quick comments, first, I _think_ the reason for this API where you have to pass in initial state to two places (to `createStore` AND `<Provider store={store} serverState={serializedStateFromServerSSR}>`) was that `createStore` comes from `redux`, while the `Provider` comes from `react-redux`. React hydration is a React concern, so the functionality for that should live in `react-redux` and therefor the `Provider`, so because we need a stable reference to the _initial_ state the API is built so you need to pass it in again there, instead of having React-specific logic in `redux`. So yes, passing it into `createStore` AND the `Provider` is correct.

Second, about streaming SSR. Streaming the result of the server render is fine as long as you don't start hydrating the markup on the client until _after_ you have hydrated the Redux store fully. The thing that might be tricky to implement in Redux is _incremental_ hydration, where parts of the state is hydrated at different times.
